observer.py

```diff
@@ -0,0 +1,106 @@
+from abc import ABC, abstractmethod
+import random
+
+
+class TopicStore:
+    def __init__(self, topics):
+        self.topics = topics
+
+    def is_topic_available(self, topic):
+        return topic in self.topics
+
+    def remove_topic(self, topic):
+        self.topics.remove(topic)
+
+
+class IObservable(ABC):
+
+    @abstractmethod
+    def add(self, observer):
+        pass
+
+    @abstractmethod
+    def remove(self, observer):
+        pass
+
+    @abstractmethod
+    def notify(self, item):
+        pass
+
+
+class IObserver(ABC):
+
+    @abstractmethod
+    def update(self, item):
+        pass
+
+
+class Observable(IObservable):
+    def __init__(self):
+        self.observers = []
+
+    def add(self, observer):
+        self.observers.append(observer)
+
+    def remove(self, observer):
+        self.observers.remove(observer)
+
+    def notify(self, item):
+        for observer in self.observers:
+            observer.update(item)
+
+
+class Teacher(Observable):
+    def __init__(self, topic_store):
+        self.topic_store = topic_store
+        super().__init__()
```

Comment:
empty `super().__init__()` is useless here. It is called only when we pass arguments

Reply:
It is required as the parent class has an init method and initializing observers list. If I don’t call super init then i would have error `'Teacher' object has no attribute 'observers'`
<img width="986" alt="Screenshot 2024-04-09 at 1 01 21 PM" src="https://github.com/AyeshaSiddique01/webscrapy_training/assets/89693571/14a30f7a-3453-4721-8e08-32a18ca7686f">